src/tbb/tcm.h

```diff
@@ -0,0 +1,173 @@
+  tcm_permit_state_t state;
+  tcm_permit_flags_t flags;
+} tcm_permit_t;
+
+// Support for permit handle
+
+typedef struct tcm_permit_rep_t* tcm_permit_handle_t;
+
+// Support for constraints
+
+typedef int32_t tcm_numa_node_t;
+typedef int32_t tcm_core_type_t;
+
+const int8_t tcm_automatic = -1;
+const int8_t tcm_any = -2;
+
+#define TCM_PERMIT_REQUEST_CONSTRAINTS_INITIALIZER {tcm_automatic, tcm_automatic, NULL, \
+                                                     tcm_automatic, tcm_automatic, tcm_automatic}
+
+typedef struct _tcm_cpu_constraints_t {
+  int32_t min_concurrency;
+  int32_t max_concurrency;
+  tcm_cpu_mask_t mask;
+  tcm_numa_node_t numa_id;
+  tcm_core_type_t core_type_id;
+  int32_t threads_per_core;
+} tcm_cpu_constraints_t;
+
+// Support for priorities
+
+enum tcm_request_priorities_t {
+  TCM_REQUEST_PRIORITY_LOW    = (INT32_MAX / 4) * 1,
+  TCM_REQUEST_PRIORITY_NORMAL = (INT32_MAX / 4) * 2,
+  TCM_REQUEST_PRIORITY_HIGH   = (INT32_MAX / 4) * 3
+};
+
+typedef int32_t tcm_request_priority_t;
+
+// Support for requests
+
+#define TCM_PERMIT_REQUEST_INITIALIZER {tcm_automatic, tcm_automatic, \
+                                         NULL, 0, TCM_REQUEST_PRIORITY_NORMAL, {}, {}}
+
+typedef struct _tcm_permit_request_t {
+  int32_t min_sw_threads;
+  int32_t max_sw_threads;
+  tcm_cpu_constraints_t* cpu_constraints;
+  uint32_t constraints_size;
+  tcm_request_priority_t priority;
+  tcm_permit_flags_t flags;
+  char reserved[4];
+} tcm_permit_request_t;
+
+// Support for client callback
+
+typedef tcm_result_t (*tcm_callback_t)(tcm_permit_handle_t p, void* callback_arg, tcm_callback_flags_t);
+
+#if WIN32
+  #define __TCM_EXPORT __declspec(dllexport)
+#else
+  #define __TCM_EXPORT
+#endif
+
+
+__TCM_EXPORT tcm_result_t tcmConnect(tcm_callback_t callback,
+                                     tcm_client_id_t *client_id);
+__TCM_EXPORT tcm_result_t tcmDisconnect(tcm_client_id_t client_id);
+
+__TCM_EXPORT tcm_result_t tcmRequestPermit(tcm_client_id_t client_id,
+                                           tcm_permit_request_t request,
+                                           void* callback_arg,
+                                           tcm_permit_handle_t* permit_handle,
+                                           tcm_permit_t* permit);
+
+__TCM_EXPORT tcm_result_t tcmGetPermitData(tcm_permit_handle_t permit_handle,
+                                           tcm_permit_t* permit);
+
+__TCM_EXPORT tcm_result_t tcmReleasePermit(tcm_permit_handle_t permit);
+
+__TCM_EXPORT tcm_result_t tcmIdlePermit(tcm_permit_handle_t permit_handle);
+
+__TCM_EXPORT tcm_result_t tcmDeactivatePermit(tcm_permit_handle_t permit_handle);
+
+__TCM_EXPORT tcm_result_t tcmActivatePermit(tcm_permit_handle_t permit_handle);
+
+__TCM_EXPORT tcm_result_t tcmRegisterThread(tcm_permit_handle_t permit_handle);
+
+__TCM_EXPORT tcm_result_t tcmUnregisterThread();
+
+__TCM_EXPORT tcm_result_t tcmGetVersionInfo(char* buffer, uint32_t buffer_size);
```

Comment:
Shouldn't we keep them under `__TBB_WEAK_SYMBOLS_PRESENT` since we only need them when weak symbols are supported?

Reply:
Not sure, it is, basically, a declaration of TCM API. Do you think it should be under this macro?

Reply:
I am not sure either, but we do that, for example here https://github.com/oneapi-src/oneTBB/blob/master/src/tbb/allocator.cpp#L54

Reply:
I actually don't have an answer for this. I would prefer keep it without macro.